src/abbrs.h

```diff
@@ -11,139 +11,25 @@
-    wcstring text{};
-
-    /// The new cursor location, or none to use the default.
-    /// This is relative to the original range.
-    maybe_t<size_t> cursor{};
-
-    /// Construct a replacement from a replacer.
-    /// The \p range is the range of the text matched by the replacer in the command line.
-    /// The text is passed in separately as it may be the output of the replacer's function.
-    static abbrs_replacement_t from(source_range_t range, wcstring text,
-                                    const abbrs_replacer_t &replacer);
+#else
+enum class abbrs_position_t : uint8_t {
+    command,
+    anywhere,
 };
 
-class abbrs_set_t {
-   public:
-    /// \return the list of replacers for an input token, in priority order.
-    /// The \p position is given to describe where the token was found.
-    abbrs_replacer_list_t match(const wcstring &token, abbrs_position_t position) const;
-
-    /// \return whether we would have at least one replacer for a given token.
-    bool has_match(const wcstring &token, abbrs_position_t position) const;
-
-    /// Add an abbreviation. Any abbreviation with the same name is replaced.
-    void add(abbreviation_t &&abbr);
+struct abbrs_replacer_t;
 
-    /// Rename an abbreviation. This asserts that the old name is used, and the new name is not; the
-    /// caller should check these beforehand with has_name().
-    void rename(const wcstring &old_name, const wcstring &new_name);
+struct abbrs_replacement_t;
 
-    /// Erase an abbreviation by name.
-    /// \return true if erased, false if not found.
-    bool erase(const wcstring &name);
+struct abbreviation_t;
 
-    /// \return true if we have an abbreviation with the given name.
-    bool has_name(const wcstring &name) const { return used_names_.count(name) > 0; }
-
-    /// \return a reference to the abbreviation list.
-    const std::vector<abbreviation_t> &list() const { return abbrs_; }
-
-    /// Import from a universal variable set.
-    void import_from_uvars(const std::unordered_map<wcstring, env_var_t> &uvars);
-
-   private:
-    /// List of abbreviations, in definition order.
-    std::vector<abbreviation_t> abbrs_{};
-
-    /// Set of used abbrevation names.
-    /// This is to avoid a linear scan when adding new abbreviations.
-    std::unordered_set<wcstring> used_names_;
-};
-
-/// \return the global mutable set of abbreviations.
-acquired_lock<abbrs_set_t> abbrs_get_set();
-
-/// \return the list of replacers for an input token, in priority order, using the global set.
-/// The \p position is given to describe where the token was found.
-inline abbrs_replacer_list_t abbrs_match(const wcstring &token, abbrs_position_t position) {
-    return abbrs_get_set()->match(token, position);
-}
+// Hacks to allow us to compile without Rust headers.
+abbrs_replacement_t abbrs_replacement_from(source_range_t range, wstring text,
+                                           wstring set_cursor_marker, bool has_cursor_marker) noexpect;
```

Comment:
Couple things:

1. There are multiple definitions of source_range_t/SourceRange. This is where the "undefined
  reference error" comes from. The implementation uses a different one than the C++ callers.
2. You declare `abbrs_replacement_from()` in a "Hacks to allow us to compile without Rust headers" section.
   This is only necessary if we need a declaration of `abbrs_replacement_from()` in a header.
   .cpp files like `reader.cpp` will include `abbrs.h` with `INCLUDE_RUST_HEADERS` defined,
   so it uses the declaration provided by cxx.
3. You take a `CxxWString` by value. I don't think this works and I've seen bad runtime breakage
   when passing by value a struct containing a `CxxWString`. I don't think Rust can properly
   dispose of a CxxWString so we should pass by reference.

I gotta run now but this diff should make everything work.
It needs some more work on cleaning up the `source_range_t`  dichotomy.

<details>

```diff
diff --git a/fish-rust/src/abbrs.rs b/fish-rust/src/abbrs.rs
index ebb1f06bd..4019a9431 100644
--- a/fish-rust/src/abbrs.rs
+++ b/fish-rust/src/abbrs.rs
@@ -13,9 +13,10 @@ use once_cell::sync::Lazy;
 pub use widestring::{Utf32Str as wstr, Utf32String as WString};
 
 use crate::abbrs::abbr_ffi::abbrs_replacer_t;
-use crate::{ffi::re::regex_t, parse_constants::SourceRange};
+use crate::ffi::re::regex_t;
+use crate::parse_constants::SourceRange;
 
-use self::abbr_ffi::{abbreviation_t, abbrs_position_t, abbrs_replacement_t, source_range_t};
+use self::abbr_ffi::{abbreviation_t, abbrs_position_t, abbrs_replacement_t};
 
 #[cxx::bridge]
 mod abbr_ffi {
@@ -23,7 +24,7 @@ mod abbr_ffi {
         include!("re.h");
         include!("parse_constants.h");
 
-        type source_range_t = crate::ffi::source_range_t;
+        type SourceRange = crate::parse_constants::SourceRange;
     }
 
     enum abbrs_position_t {
@@ -39,7 +40,7 @@ mod abbr_ffi {
     }
 
     struct abbrs_replacement_t {
-        range: source_range_t,
+        range: SourceRange,
         text: UniquePtr<CxxWString>,
         cursor: usize,
         has_cursor: bool,
@@ -66,9 +67,9 @@ mod abbr_ffi {
 
         #[cxx_name = "abbrs_replacement_from"]
         fn abbrs_replacement_from_ffi(
-            range: source_range_t,
-            text: CxxWString,
-            set_cursor_marker: CxxWString,
+            range: SourceRange,
+            text: &CxxWString,
+            set_cursor_marker: &CxxWString,
             has_cursor_marker: bool,
         ) -> abbrs_replacement_t;
 
@@ -253,9 +254,9 @@ impl Replacement {
 }
 
 fn abbrs_replacement_from_ffi(
-    range: source_range_t,
-    text: CxxWString,
-    set_cursor_marker: CxxWString,
+    range: SourceRange,
+    text: &CxxWString,
+    set_cursor_marker: &CxxWString,
     has_cursor_marker: bool,
 ) -> abbrs_replacement_t {
     let mut cursor_marker = None;
diff --git a/fish-rust/src/ffi.rs b/fish-rust/src/ffi.rs
index ac2a37ba2..d61341f33 100644
--- a/fish-rust/src/ffi.rs
+++ b/fish-rust/src/ffi.rs
@@ -32,7 +32,7 @@ include_cpp! {
     safety!(unsafe_ffi)
 
     generate_pod!("wcharz_t")
-    generate_pod!("source_range_t")
+    // generate_pod!("source_range_t")
     generate!("make_fd_nonblocking")
     generate!("wperror")
 
diff --git a/src/abbrs.h b/src/abbrs.h
index 8bd479dec..3eeb50024 100644
--- a/src/abbrs.h
+++ b/src/abbrs.h
@@ -27,9 +27,6 @@ struct abbrs_replacement_t;
 
 struct abbreviation_t;
 
-// Hacks to allow us to compile without Rust headers.
-abbrs_replacement_t abbrs_replacement_from(source_range_t range, wstring text,
-                                           wstring set_cursor_marker, bool has_cursor_marker) noexpect;
 #endif
 
 #endif
diff --git a/src/parse_constants.h b/src/parse_constants.h
index a7c3e75e6..a6e12fc5e 100644
--- a/src/parse_constants.h
+++ b/src/parse_constants.h
@@ -31,10 +31,11 @@ using parse_error_list_t = ParseErrorList;
 
 #include "config.h"
 
-struct source_range_t {
+struct SourceRange {
     source_offset_t start;
     source_offset_t length;
 };
+using source_range_t = SourceRange;
 
 enum class parse_token_type_t : uint8_t {
     invalid = 1,
diff --git a/src/reader.cpp b/src/reader.cpp
index 089edcfd1..f9a0e3f2f 100644
--- a/src/reader.cpp
+++ b/src/reader.cpp
@@ -1376,13 +1376,14 @@ void reader_data_t::pager_selection_changed() {
 
 /// Expand an abbreviation replacer, which may mean running its function.
 /// \return the replacement, or none to skip it. This may run fish script!
-maybe_t<abbrs_replacement_t> expand_replacer(source_range_t range, const wcstring &token,
+maybe_t<abbrs_replacement_t> expand_replacer(SourceRange range, const wcstring &token,
                                              const abbrs_replacer_t &repl, parser_t &parser) {
     if (!repl.is_function) {
         // Literal replacement cannot fail.
         FLOGF(abbrs, L"Expanded literal abbreviation <%ls> -> <%ls>", token.c_str(),
               (*repl.replacement).c_str());
-        return abbrs_replacement_from(range, *repl.replacement, *repl.set_cursor_marker, repl.has_cursor_marker);
+        return abbrs_replacement_from(range, *repl.replacement, *repl.set_cursor_marker,
+                                      repl.has_cursor_marker);
     }
 
     wcstring cmd = escape_string(*repl.replacement);
@@ -1398,8 +1399,7 @@ maybe_t<abbrs_replacement_t> expand_replacer(source_range_t range, const wcstrin
     }
     wcstring result = join_strings(outputs, L'\n');
     FLOGF(abbrs, L"Expanded function abbreviation <%ls> -> <%ls>", token.c_str(), result.c_str());
-    return abbrs_replacement_from(range, std::move(result), *repl.set_cursor_marker,
-                                  repl.has_cursor_marker);
+    return abbrs_replacement_from(range, result, *repl.set_cursor_marker, repl.has_cursor_marker);
 }
 
 // Extract all the token ranges in \p str, along with whether they are an undecorated command.
```

</details>

Reply:
Thank you very much, Will try this. I have no experience in CPP so just trying to get it to compile first and then fix any issues